lib/subgraphManagers/IntrospectAddNodeAndCompose.ts

```diff
@@ -0,0 +1,321 @@
+import type {
+  DocumentNode,
+  GraphQLSchema,
+  ObjectTypeDefinitionNode,
+} from 'graphql';
+import { visit } from 'graphql';
+import gql from 'graphql-tag';
+
+import resolvable from '@josephg/resolvable';
+
+import type {
+  GraphQLDataSource,
+  ServiceDefinition,
+  ServiceEndpointDefinition,
+  SubgraphHealthCheckFunction,
+  SupergraphManager,
+  SupergraphSdlUpdateFunction,
+} from '@apollo/gateway';
+import type { SupergraphSdlHookOptions } from '@apollo/gateway/dist/config';
+import type { IntrospectAndComposeOptions } from '@apollo/gateway/dist/supergraphManagers/IntrospectAndCompose';
+import { LocalGraphQLDataSource } from '@apollo/gateway';
+import { composeServices } from '@apollo/composition';
+import {
+  loadServicesFromRemoteEndpoint,
+  Service,
+} from '@apollo/gateway/dist/supergraphManagers/IntrospectAndCompose/loadServicesFromRemoteEndpoint';
+
+import TypeIDDataSource from 'lib/utils/TypeIDDataSource';
+
+import type { ExternalDirectives } from '../service';
+import createNodeService from '../service';
+import { NODE_SERVICE_NAME, NODE_SERVICE_URL } from 'lib/utils/constants';
+import NoInitialSdlError from 'lib/errors/NoInitialSdlError';
+import NoSubgraphsError from 'lib/errors/NoSubgraphsError';
+import NoGetDataSourceMethodDefinedError from 'lib/errors/NoGetDataSourceMethodDefinedError';
+import PollWithoutConfiguredIntervalError from 'lib/errors/PollWithoutConfiguredIntervalError';
+import SupergraphCompositionError from 'lib/errors/SupergraphCompositionError';
+
+export type NodeGatewayConfig = {
+  nodeTypeDefs?: DocumentNode[];
+  typeIDDataSource: TypeIDDataSource;
+  directives?: ExternalDirectives;
+  nodeQueryDirectives?: string | null;
+};
+
+type State =
+  | { phase: 'initialized' }
+  | { phase: 'polling'; pollingPromise?: Promise<void> }
+  | { phase: 'stopped' };
+
+type NodeTypes = {
+  nodeTypeDefs: DocumentNode[];
+  seenNodeTypes: Set<string>;
+  serviceDefinitions: ServiceDefinition[];
+};
+
+type GetDataSourceWithSchema = (
+  serviceDefinition: ServiceEndpointDefinition & { schema?: GraphQLSchema },
+) => GraphQLDataSource;
+
+export default class IntrospectAddNodeAndCompose implements SupergraphManager {
+  private static objectImplementsNodeInterface = (
+    node: ObjectTypeDefinitionNode,
+  ): boolean =>
+    node.interfaces?.some(({ name }): boolean => name.value === 'Node') ||
+    false;
+
+  private static generateNodeType = (name: string): DocumentNode =>
+    gql`
+    extend type ${name} implements Node @key(fields: "id") {
+      id: ID! @external
+    }
+  `;
```

Comment:
both of these `private static` could be toplevel functions, or moved to a utils file... no real need or use to be in the class.